rust/lancedb/src/database.rs

```diff
@@ -0,0 +1,109 @@
+//! and is based off listing directories in a filesystem.
+//!
+//! Users may want to provider their own implementations for a variety of reasons:
+//!  * Tables may be arranged in a different order on the S3 filesystem
+//!  * Tables may be managed by some kind of independent application (e.g. some database)
+//!  * Tables may be managed by a database system (e.g. Postgres)
+//!  * A custom table implementation (e.g. remote table, etc.) may be used
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use arrow_array::RecordBatchReader;
+use lance::dataset::ReadParams;
+
+use crate::error::Result;
+use crate::table::{TableDefinition, TableInternal, WriteOptions};
+
+pub mod listing;
+
+pub trait DatabaseOptions {
+    fn serialize_into_map(&self, map: &mut HashMap<String, String>);
+}
+
+/// A request to list names of tables in the database
+#[derive(Clone, Debug, Default)]
+pub struct TableNamesRequest {
+    /// If present, only return names that come lexicographically after the supplied
+    /// value.
+    ///
+    /// This can be combined with limit to implement pagination by setting this to
+    /// the last table name from the previous page.
+    pub start_after: Option<String>,
+    /// The maximum number of table names to return
+    pub limit: Option<u32>,
+}
+
+/// A request to open a table
+#[derive(Clone, Debug)]
+pub struct OpenTableRequest {
+    pub name: String,
+    pub index_cache_size: Option<u32>,
+    pub lance_read_params: Option<ReadParams>,
+}
+
+pub type TableBuilderCallback = Box<dyn FnOnce(OpenTableRequest) -> OpenTableRequest + Send>;
+
+/// Describes what happens when creating a table and a table with
+/// the same name already exists
+pub enum CreateTableMode {
+    /// If the table already exists, an error is returned
+    Create,
+    /// If the table already exists, it is opened.  Any provided data is
+    /// ignored.  The function will be passed an OpenTableBuilder to customize
+    /// how the table is opened
+    ExistOk(TableBuilderCallback),
+    /// If the table already exists, it is overwritten
+    Overwrite,
+}
+
+impl CreateTableMode {
+    pub fn exist_ok(
+        callback: impl FnOnce(OpenTableRequest) -> OpenTableRequest + Send + 'static,
+    ) -> Self {
+        Self::ExistOk(Box::new(callback))
+    }
+}
+
+impl Default for CreateTableMode {
+    fn default() -> Self {
+        Self::Create
+    }
+}
+
+/// A request to create a table
+#[derive(Default)]
+pub struct CreateTableRequest {
+    pub name: String,
+    pub data: Option<Box<dyn RecordBatchReader + Send>>,
+    pub mode: CreateTableMode,
+    pub write_options: WriteOptions,
+    pub table_definition: Option<TableDefinition>,
+}
+
+/// The `Database` trait defines the interface for database implementations.
+///
+/// A database is responsible for managing tables and their metadata.
+#[async_trait::async_trait]
+pub trait Database: Send + Sync + std::fmt::Debug + std::fmt::Display + 'static {
```

Comment:
I wonder if we want to have them all implement `std::any::Any`, so we could do downcasting? Or maybe an `as_any()` as is done in DataFusion. (I forget why they do that...)

Reply:
Fair enough.  I've added `std::any::Any` and `as_any` (I don't know exactly why but I think `as_any` is a required part of the pattern).